src/psyclone/psyad/tl2ad.py

```diff
@@ -98,9 +155,407 @@ def generate_adjoint(tl_psyir):
+        statements.append(
+            Assignment.create(Reference(sym_copy), Reference(sym)))
+
+    # Call the TL kernel
+    statements.append(Call.create(tl_kernel_sym,
+                                  [Reference(sym) for sym in new_arg_list]))
+
+    # Compute the inner product of the result of the TL kernel
+    statements += _create_inner_product(inner1,
+                                        [(sym, sym) for sym in inputs])
+
+    # Call the adjoint kernel using the outputs of the TL kernel as input
+    statements.append(Call.create(adj_kernel_sym,
+                                  [Reference(sym) for sym in new_arg_list]))
+
+    # Compute inner product of result of adjoint kernel with original inputs
+    statements += _create_inner_product(inner2,
+                                        zip(inputs, input_copies))
+
+    # Compare the inner products
+    tol_zero = Literal("1.0e-10", REAL_DOUBLE_TYPE)
+
+    diff = BinaryOperation.create(BinaryOperation.Operator.SUB,
+                                  Reference(inner1), Reference(inner2))
+    abs_diff = UnaryOperation.create(UnaryOperation.Operator.ABS, diff)
+    statements.append(Assignment.create(Reference(diff_sym), abs_diff))
+
+    # If the test fails then the harness will print a message and return early.
+    ptree = Fortran2003.Write_Stmt(
+        "write(*,*) 'Test of adjoint of ''{0}'' failed: diff = ', {1}".format(
+            tl_kernel.name, diff_sym.name))
+
+    statements.append(
+        IfBlock.create(BinaryOperation.create(BinaryOperation.Operator.GT,
+                                              Reference(diff_sym),
+                                              tol_zero.copy()),
+                       [CodeBlock([ptree], CodeBlock.Structure.STATEMENT),
+                        Return()]))
+
+    # Otherwise the harness prints a message reporting that all is well.
+    ptree = Fortran2003.Write_Stmt(
+        "write(*,*) 'Test of adjoint of ''{0}'' passed: diff = ', {1}".format(
+            tl_kernel.name, diff_sym.name))
+    statements.append(CodeBlock([ptree], CodeBlock.Structure.STATEMENT))
+
+    # Finally, create driver program from the list of statements.
+    routine = Routine.create(
+        "adj_test", symbol_table, statements, is_program=True)
+
+    return routine
+
+
+def _create_inner_product(result, symbol_pairs):
+    '''
+    Creates PSyIR that computes the inner product of each pair of symbols
+    in the supplied list and accumulates it into the supplied `result`
+    variable (which is first zeroed).
+
+    :param result: symbol which will accumulate result.
+    :type result: :py:class:`psyclone.psyir.symbols.DataSymbol`
+    :param symbol_pairs: list of pairs of symbols for which to compute inner \
+                         products.
+    :type symbol_pairs: list of 2-tuples of \
+                        :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+    :returns: PSyIR that performs the inner product and accumulates the result\
```

Comment:
Picky but I always leave a space before a '\'

Reply:
I try unless there isn't room. However, since it's you I've moved the "result" down to the next line :-)